tests/prefetch_related/tests.py

```diff
@@ -2073,3 +2074,197 @@ def test_empty_order(self):
             with self.subTest(book=book):
                 self.assertEqual(len(book.authors_sliced), 1)
                 self.assertIn(book.authors_sliced[0], list(book.authors.all()))
+
+
+class PrefetchRelatedCacheWorksWithInheritance(TestCase):
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.book = Book.objects.create(title="Book1")
+        cls.related1 = Author.objects.create(name="related1", first_book=cls.book)
+        cls.related2 = Author.objects.create(name="related2", first_book=cls.book)
+        cls.related3 = Author.objects.create(name="related3", first_book=cls.book)
+        cls.related4 = Author.objects.create(name="related4", first_book=cls.book)
+
+        cls.child = AuthorWithAgeChild.objects.create(
+            name="child",
+            age=31,
+            first_book=cls.book,
+        )
+        cls.m2m_child = AuthorWithAgeChild.objects.create(
+            name="m2m_child",
+            age=31,
+            first_book=cls.book,
+        )
+        cls.m2m_child.favorite_authors.set([cls.related1, cls.related2, cls.related3])
+
+    def test_parent_fk_available_in_child(self):
+        """
+        Make sure a prefetched FK relation is still cached in a direct child
+        """
+        qs = (
+            Author.objects.select_related("authorwithage")
+            .prefetch_related("first_book")
+            .filter(pk=self.child.pk)
+        )
+        with self.assertNumQueries(2):
+            results = list(qs)
+            self.assertEqual(len(results), 1)
+            self.assertEqual(results[0].authorwithage.first_book, self.book)
+
+    def test_grandparent_fk_available_in_child(self):
+        """
+        Make sure a prefetched FK relation is still cached in a grandchild
+        """
+        qs = (
+            Author.objects.select_related(
+                "authorwithage", "authorwithage__authorwithagechild"
+            )
+            .prefetch_related("first_book")
+            .filter(pk=self.child.pk)
+        )
+        with self.assertNumQueries(2):
+            results = list(qs)
+            self.assertEqual(len(results), 1)
+            self.assertEqual(
+                results[0].authorwithage.authorwithagechild.first_book, self.book
+            )
```

Comment:
Do we know if the other way around works as well?

e.g.

```python
qs = (
    AuthorWithAgeChild.objects
    .prefetch_related("first_book")
    .filter(pk=self.child.pk)
)
with self.assertNumQueries(2):
    results = list(qs)
    self.assertEqual(len(results), 1)
    self.assertEqual(
        results[0].first_book, self.book
    )
    self.assertEqual(
        results[0].authorwithage_ptr.first_book, self.book
    )
    self.assertEqual(
        results[0].authorwithage_ptr.author_ptr.first_book, self.book
    )
```

Reply:
It does not, the code only looks for the cache in ancestors, not descendants.
Do you think support for that should be added here, too? If so, I'd appreciate some pointers for how to do so, I am not aware of an API for "give me all pointers to descendants" in `Options`, which would be the API to use, similar to `get_ancestor_link` for walking "up" to ancestors.

Reply:
I think it's fine for now!

I assume this breaks because `prefetch_related` could not take advantage of the logic that you just added to _walk_ up to the ancestor when caching the field from prefetched data. In order to make it work I suspect we'd have to _teach_ it to cache the value in the proper location but then it open a new can of worm when doing things like

```python
AuthorWithAge.objects.only("age").prefetch_related("first_book")
```

As normally `author_with_age.author_ptr` would not be automatically created under these circumstances and thus we couldn't store the prefetched `first_book` on `author_with_age.author_ptr._prefetch_related_cache` so it can be retrieved by doing `author_with_age.author_ptr.first_book`.

Reply:
I see what you mean now! I didn't even consider the option of storing the cache in a different location in the first place.
But as you said, that is a _big_ can of worms, because then you'd have different behavior of the prefetched objects based on whether other things are cached or not.